Verify rpc request user against client cert

mberhault opened the pull request:
I would like a bit of feedback on the reflection in rpc/codec/server.go:userFromRequest.
Believe or not, this is the first time I'm using the reflect package.

Of course, comments on the rest of the code are more than welcome, keeping in mind that this is still a work in progress.

Commit:
Check rpc request.Header.User against client certs

All requests now implement UserRequest (func GetUser() string).
In secure more, make sure the client certificate user matches
the request user, or "node".

kkaneda committed:
Mention required git version info (1.8 or above) in CONTRIBUTING.md

Fix #1409

tschottdorf commented on rpc/codec/server.go:
```diff
@@ -79,6 +92,72 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 
+// userFromRequest takes a request proto and attemps to access
+// request.RequestHeader.User.
+func (c *serverCodec) userFromRequest(request proto.Message) (string, error) {
```
can this method not take a `request proto.Request`?

MBerhault replied:
removed function now that "extract user from header" is just 3 lines

tschottdorf commented on rpc/codec/server.go:
```diff
@@ -79,6 +92,72 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
+	case *message.EchoRequest, *message.ArithRequest:
+		// These are for testing purposes only.
+		return nil
+	default:
```
s/default/proto.Request/
and then a `default` case which errors out.

MBerhault replied:
oh yeah, much better. I wasn't aware of proto.Request.
This simplifies everything.

tschottdorf commented on rpc/codec/server.go:
```diff
@@ -24,11 +24,17 @@ package codec
 	"net/rpc"
 
+	"github.com/cockroachdb/cockroach/base"
+	cockroach_proto "github.com/cockroachdb/cockroach/proto"
```
doesn't really concern your PR, but I've noticed how `etcd/raft` named its proto package `raftpb`, so if it's not mandated by style to have our proto package called `proto`, we could just rename both the `proto` folder and the contained package to `roachpb` (or even shorter, `pb`), which would allow us to always use the canonical import path.

MBerhault replied:
that would be much better, but yes, this is rather outside the scope of this PR.

tschottdorf replied:
absolutely. @petermattis?

petermattis replied:
We (err, @tamird) is looking into splitting up the `proto` package. There is no particular reason to have the protos separated from their usage. For example, the `MVCC` protos should live in `storage/engine`. We might have a base protos package which would make sense to name `roachpb` or `pb`, but hold off on doing any renaming yet until @tamird gets started with the split.

tschottdorf commented on rpc/codec/server.go:
```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
+		}
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest, *cockroach_proto.RaftMessageRequest:
+		// Ping and Gossip requests do not have a header: require the node user.
+		// TODO(marc): it may be worth it to make them implement Request.
```
yeah, probably worth having everything implement that to allow you to be agnostic here and below.

MBerhault replied:
yup. I started working on that, but ran into protoc issues.
At the very least, I want to remove all test requests.

petermattis commented on rpc/codec/server.go:
```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
+		if len(requestedUser) == 0 {
+			return util.Errorf("missing User in request header: %+v", header)
+		}
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest, *cockroach_proto.RaftMessageRequest:
```
Mentioning specify protos feels awkward to me. Why not just have everything that isn't a `cockroach_proto.Request` default to not authenticated?

MBerhault replied:
until Ping/Gossip/Raft have a RequestHeader, I need to have them default to "desired user = node". otherwise, we would just fail all those requests.

petermattis commented on rpc/codec/server.go:
```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
+	var requestedUser string
+
+	switch request := request.(type) {
+	case cockroach_proto.Request:
```
Rather than use `cockroach_proto.Request` here, I think you could define your own more restrictive interface:

```
type userRequest interface {
  GetUser() string
}
```

Then you could fix the `{Echo,Arith}Request` messages by adding a user field. Ditto for `{Ping,Gossip,RaftMessage}Request`.

MBerhault replied:
Good point. I'll settle for that in the short term. Long-term, we need to see if there are more things we may need from the request header.

bdarnell commented on rpc/client_test.go:
```diff
@@ -36,13 +36,14 @@ func init() {
 func TestClientHeartbeat(t *testing.T) {
 	addr := util.CreateTestAddr("tcp")
 
+	// Heartbeats are node-to-node requests: use server context for both server and client.
```
Maybe server context should be named node context instead, since it can be used as a client too.

MBerhault replied:
Ah, yes. I was pondering how to make those clearer. I'll do it in a separate PR since we have TestServerContext all over the place.

bdarnell commented on rpc/codec/message/arith.proto:
```diff
@@ -9,8 +9,9 @@ option go_package = "message";
 message ArithRequest {
-	optional int32 a = 1 [(gogoproto.nullable) = false];
-	optional int32 b = 2 [(gogoproto.nullable) = false];
+  optional string user = 1 [(gogoproto.nullable) = false];
```
Indentation is funky here, at least in github's diff view. Mixing tabs and spaces?

MBerhault replied:
ah yes. how I wish I had gofmt for everything. Fixed.

bdarnell commented on rpc/codec/message/echo.proto:
```diff
@@ -13,7 +13,8 @@ option (gogoproto.marshaler_all) = true;
 
 message EchoRequest {
-	optional string msg = 1 [(gogoproto.nullable) = false];
+  optional string user = 1 [(gogoproto.nullable) = false];
```
Indention is off here too.

MBerhault replied:
Done.

petermattis commented on proto/api.go:
```diff
@@ -86,6 +86,12 @@ type Request interface {
 }
 
+// RequestWithUser is an interface for RPC requests that have a "requested user".
+type RequestWithUser interface {
```
`UserRequest`? Note that there is no need to define this interface here. You could define it in rpc/codec right next to its use. (And name it `userRequest` or something like that).

MBerhault replied:
I thought about that but figured it would be clearer to have it in the same place as Request and its implementers. I'm happy moving it though if you insist.

petermattis replied:
Moving it removes the dependency on the `proto` package. I like that.

MBerhault replied:
Good point. Will move it along with the rest of the code into a hook someplace (most likely security/)

petermattis commented on rpc/codec/server.go:
```diff
@@ -79,6 +90,43 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
+// We compare the header.User against the client certificate Subject.CommonName.
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
+	// RequestWithUser must be implemented.
+	requestWithUser, ok := request.(cockroach_proto.RequestWithUser)
```
As mentioned above, you can just define the interface here. Within the `authenticateRequest` function even:

```
  type requestWithUser interface {
    GetUser() string
  }
```

petermattis commented on rpc/codec/server.go:
```diff
@@ -79,6 +90,43 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// Requests need to implement RequestWithUser.
+// We compare the header.User against the client certificate Subject.CommonName.
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
```
Rather than performing the authentication here, perhaps serverCodec should contain a `PreHook` that could be invoked on a message before it is dispatched. That would keep the `rpc/codec` package free of dependencies of other cockroach packages.

MBerhault replied:
I haven't changed it to a hook yet, just moved the certificate check to connection time (which is the right thing to do either way).
code still depends on security just to check user vs NodeUser (this should probably move).
I'll try to add the hook such that we can use it from the http server as well.

petermattis commented:
LGTM, though I think a small change could reduce dependency creep.

bdarnell commented:
LGTM

petermattis commented on rpc/codec/message/arith.proto:
```diff
@@ -9,8 +9,9 @@ option go_package = "message";
 message ArithRequest {
-	optional int32 a = 1 [(gogoproto.nullable) = false];
-	optional int32 b = 2 [(gogoproto.nullable) = false];
+	optional string user = 1 [(gogoproto.nullable) = false];
```
Similar to `GossipRequest`, you could add an `arith.go` package and manually implemented `ArithRequest.GetUser()` method.

MBerhault replied:
Done.

petermattis commented on rpc/codec/server.go:
```diff
@@ -79,6 +87,49 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 }
 
+// UserRequest is an interface for RPC requests that have a "requested user".
+type UserRequest interface {
```
No need to export this type.

MBerhault replied:
This has moved to security/tls.go, inside the authentication method and unexported.

petermattis replied:
Looks like you didn't delete `UserRequest` here.

spencerkimball commented:
Nice.

Commit:
Use security.AuthenticationHook to handle insecure/certs/NodeUser logic

Commit:
Use security.AuthenticationHook to handle insecure/certs/NodeUser logic

MBerhault commented on rpc/server.go:
```diff
@@ -88,24 +88,28 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
+
+	// Construct an authentication hook for this security mode and TLS state.
+	authHook, err := security.AuthenticationHook(s.context.Insecure, r.TLS)
```
@petermattis: how about something like this?

petermattis replied:
Yep. Looks good.

petermattis committed:
Fix go linking error by providing a dummy file.

bdarnell committed:
Remove 'cover' from GLOCKFILE (partially undoing e7481ded650).

'cover', like 'vet', is normally installed by go itself and is
special-cased in 'go get' to install under GOROOT instead of GOPATH.
The user may not have write access to GOROOT, so we should just rely on
'cover' being installed by go instead of installing it with glock.

bdarnell committed:
Merge pull request #1420 from bdarnell/cover

Remove 'cover' from GLOCKFILE (partially undoing e7481ded650).

tamird committed:
Merge pull request #1419 from cockroachdb/pmattis/structured

Fix go linking error by providing a dummy file.

tamird committed:
Merge pull request #1421 from tamird/mvcc-proto

Split out MVCC protos into storage/engine

petermattis commented on rpc/codec/server.go:
```diff
@@ -37,6 +38,9 @@ type serverCodec struct {
 	methods []string
 
+	// Authentication hook. May be nil in tests.
+	authenticationHook func(string) error
```
The hook doesn't need to be authentication specific. Make the function take a `proto.Message` and have the proto perform a type assertion or type switch to extract the user.

MBerhault replied:
Made the hook take the entire proto.Message (see security/tls.go)

kkaneda committed:
Merge pull request #1412 from cockroachdb/kkaneda/git_version

Mention required git version info (1.8 or above) in CONTRIBUTING.md

Commit:
Make authentication hook take a proto.Message

Commit:
Make authentication hook take a proto.Message

Commit:
Implement userRequest in test protos

Commit:
Implement userRequest in test protos

Commit:
Authentication hook test and a bit of refactoring

Commit:
Authentication hook test and a bit of refactoring

Commit:
Add root test, fix typo

Commit:
Add root test, fix typo

Commit:
Resolve merge conflicts

Commit:
Merge branch 'marc/verify_rpc_user' of github.com:cockroachdb/cockroach into marc/verify_rpc_user

Conflicts:
	gossip/client_test.go
	rpc/client_test.go
	rpc/send_test.go
	rpc/server.go

petermattis commented on rpc/codec/rpc_test.go:
```diff
@@ -112,7 +112,7 @@ func listenAndServeArithAndEchoService(network, addr string) (net.Addr, error) {
 				continue
 			}
-			go srv.ServeCodec(NewServerCodec(conn))
+			go srv.ServeCodec(NewServerCodec(conn, nil /* no authentication hook */))
```
These comments rotted: `s/authentication hook/request hook/g`

MBerhault replied:
separate PR on its way to fix that and extraneous UserRequest.
